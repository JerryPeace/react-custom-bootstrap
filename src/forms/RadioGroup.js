import PropTypes from 'prop-types';
import React, { Component } from 'react';
import classNames from 'classnames';
import Radio from './Radio';
import LayoutHelper from './LayoutHelper';

// There are two methods to handle 'checked' to radio in RadioGroup
// 1. use selectedValue
// 2. set value and cehcekd in each item in 'items'
class RadioGroup extends Component {

    static propTypes = {
        inline: PropTypes.bool,
        items: PropTypes.array,
        labelAlign: PropTypes.string,
        selectedValue: PropTypes.any,
        standalone: PropTypes.bool,
        onChange: PropTypes.func,
        wrapperClassName: PropTypes.string,
        labelText: PropTypes.node
    };

    static defaultProps = {
        bsClass: 'radio-group',
        items: [],
        inline: false,
        labelAlign: 'right',
        standalone: false
    };

    constructor(props) {
        super(props);
        this.radios = [];
    }

    isDirty() {
        for (let i = 0; i < this.radios.length; i++) {
            let childCom = this.radios[i];
            if (childCom.isDirty && childCom.isDirty()) {
                return true;
            }
        }
        return false;
    }

    cleanDirty() {
        this.radios.map((childEl) => {
            // let childCom = this.refs[childEl.ref];
            childEl.cleanDirty && childEl.cleanDirty();
        });
    }

    getValue() {
        let value;

        for (let i = 0; i < this.radios.length; i++) {
            let childCom = this.radios[i];
            let checked = childCom.getChecked();

            if (checked) {
                return childCom.getValue();
            }
        }
        return null;
    }

    renderInputGroup() {
        const {
            items,
            inline,
            onChange,
            selectedValue,
            ...restProps
        } = this.props;

        this.childrenEls = items.map((item, idx) => {
            return (
                <Radio
                    checked={selectedValue === item.value}
                    inline={inline}
                    key={idx}
                    onChange={onChange}
                    ref={n => this.radios[idx] = n}
                    standalone={true}
                    {...item}
                />
            );
        });

        return this.childrenEls;
    }

    render() {
        const wrapperClassName = this.props.wrapperClassName || classNames('col-xs-10', {
            'col-xs-offset-2': !this.props.labelText
        });

        return (
            <LayoutHelper
                {...this.props}
                FieldBase={this.renderInputGroup()}
                wrapperClassName={wrapperClassName} />
        );
    }
}

export default RadioGroup;
