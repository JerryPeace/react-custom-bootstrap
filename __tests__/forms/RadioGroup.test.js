import React from 'react';
import { mount } from 'enzyme';
import toJSON from'enzyme-to-json';
import RadioGroup from 'forms/RadioGroup';

describe('RadioGroup', () => {
    let mockChange;
    let items;
    let radios;

    describe('with props', () => {
        beforeEach(() => {
            mockChange = jest.fn();
            items = [
                {
                    labelText: 'apple',
                    value: 1
                }, {
                    labelText: 'pen',
                    value: 2
                }, {
                    labelText: 'pine apple',
                    value: 3
                }
            ];
            radios = mount(
                <RadioGroup
                    items={items}
                    onChange={mockChange}
                    selectedValue={1}
                    />
            );
        });

        test('should render corrently', () => {
            expect(toJSON(radios)).toMatchSnapshot();
            radios.instance().cleanDirty();
            expect(radios.instance().isDirty()).toBeFalsy();
            expect(radios.instance().getValue()).toBe('1');
        });

        test('check isDirty revert behavior', () => {
            radios.instance().cleanDirty();
            expect(radios.instance().isDirty()).toBeFalsy();
            radios.setProps({selectedValue: 2});
            expect(radios.instance().isDirty()).toBeTruthy();
            radios.setProps({selectedValue: 1});
            expect(radios.instance().isDirty()).toBeFalsy();
        });

        test('click input 2 would trigger callback', () => {
            radios.find({value: 2}).simulate('change');
            expect(mockChange.mock.calls.length).toBe(1);
            expect(mockChange.mock.calls[0][1]).toBe(2);
        });

        test('render correctly when selected changed', () => {
            radios.setProps({selectedValue: 2});
            expect(radios.instance().getValue()).toBe('2');
            expect(toJSON(radios)).toMatchSnapshot();
        });
    });
});
