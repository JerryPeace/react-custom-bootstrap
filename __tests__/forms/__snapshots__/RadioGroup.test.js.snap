// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup with props render correctly when selected changed 1`] = `
<RadioGroup
  bsClass="radio-group"
  inline={false}
  items={
    Array [
      Object {
        "labelText": "apple",
        "value": 1,
      },
      Object {
        "labelText": "pen",
        "value": 2,
      },
      Object {
        "labelText": "pine apple",
        "value": 3,
      },
    ]
  }
  labelAlign="right"
  onChange={[Function]}
  selectedValue={2}
  standalone={false}
>
  <LayoutHelper
    FieldBase={
      Array [
        <Radio
          bsClass="radio"
          checked={false}
          inline={false}
          inputType="radio"
          labelAlign="right"
          labelText="apple"
          onChange={[Function]}
          standalone={true}
          value={1}
          wrapperClassName="col-xs-offset-2 col-xs-10"
      />,
        <Radio
          bsClass="radio"
          checked={true}
          inline={false}
          inputType="radio"
          labelAlign="right"
          labelText="pen"
          onChange={[Function]}
          standalone={true}
          value={2}
          wrapperClassName="col-xs-offset-2 col-xs-10"
      />,
        <Radio
          bsClass="radio"
          checked={false}
          inline={false}
          inputType="radio"
          labelAlign="right"
          labelText="pine apple"
          onChange={[Function]}
          standalone={true}
          value={3}
          wrapperClassName="col-xs-offset-2 col-xs-10"
      />,
      ]
    }
    bsClass="radio-group"
    inline={false}
    items={
      Array [
        Object {
          "labelText": "apple",
          "value": 1,
        },
        Object {
          "labelText": "pen",
          "value": 2,
        },
        Object {
          "labelText": "pine apple",
          "value": 3,
        },
      ]
    }
    labelAlign="right"
    onChange={[Function]}
    selectedValue={2}
    singleCheckboxRadio={false}
    standalone={false}
    wrapperClassName="col-xs-10 col-xs-offset-2"
  >
    <FormGroup
      className=""
    >
      <div
        className="form-group"
      >
        <div
          className="col-xs-10 col-xs-offset-2"
        >
          <div
            className="radio-group"
          >
            <Radio
              bsClass="radio"
              checked={false}
              inline={false}
              inputType="radio"
              labelAlign="right"
              labelText="apple"
              onChange={[Function]}
              standalone={true}
              value={1}
              wrapperClassName="col-xs-offset-2 col-xs-10"
            >
              <LayoutHelper
                FieldBase={
                  <input
                    checked={false}
                    onChange={[Function]}
                    type="radio"
                    value={1}
                  />
                }
                bsClass="radio"
                checked={false}
                inline={false}
                inputType="radio"
                labelAlign="right"
                labelText="apple"
                onChange={[Function]}
                singleCheckboxRadio={false}
                standalone={true}
                value={1}
                wrapperClassName="col-xs-offset-2 col-xs-10"
              >
                <div
                  className="radio"
                >
                  <Label
                    FieldBase={
                      <input
                        checked={false}
                        onChange={[Function]}
                        type="radio"
                        value={1}
                      />
                    }
                    className="checkbox-radio-label radio-wrapper"
                    inline={false}
                    inputType="radio"
                    labelAlign="right"
                    onChange={[Function]}
                    text="apple"
                    value={1}
                    wrapperClassName="col-xs-offset-2 col-xs-10"
                  >
                    <label
                      className="checkbox-radio-label radio-wrapper"
                    >
                      <input
                        checked={false}
                        onChange={[Function]}
                        type="radio"
                        value={1}
                      />
                      apple
                    </label>
                  </Label>
                </div>
              </LayoutHelper>
            </Radio>
          </div>
          <div
            className="radio-group"
          >
            <Radio
              bsClass="radio"
              checked={true}
              inline={false}
              inputType="radio"
              labelAlign="right"
              labelText="pen"
              onChange={[Function]}
              standalone={true}
              value={2}
              wrapperClassName="col-xs-offset-2 col-xs-10"
            >
              <LayoutHelper
                FieldBase={
                  <input
                    checked={true}
                    onChange={[Function]}
                    type="radio"
                    value={2}
                  />
                }
                bsClass="radio"
                checked={true}
                inline={false}
                inputType="radio"
                labelAlign="right"
                labelText="pen"
                onChange={[Function]}
                singleCheckboxRadio={false}
                standalone={true}
                value={2}
                wrapperClassName="col-xs-offset-2 col-xs-10"
              >
                <div
                  className="radio"
                >
                  <Label
                    FieldBase={
                      <input
                        checked={true}
                        onChange={[Function]}
                        type="radio"
                        value={2}
                      />
                    }
                    className="checked checkbox-radio-label radio-wrapper"
                    inline={false}
                    inputType="radio"
                    labelAlign="right"
                    onChange={[Function]}
                    text="pen"
                    value={2}
                    wrapperClassName="col-xs-offset-2 col-xs-10"
                  >
                    <label
                      className="checked checkbox-radio-label radio-wrapper"
                    >
                      <input
                        checked={true}
                        onChange={[Function]}
                        type="radio"
                        value={2}
                      />
                      pen
                    </label>
                  </Label>
                </div>
              </LayoutHelper>
            </Radio>
          </div>
          <div
            className="radio-group"
          >
            <Radio
              bsClass="radio"
              checked={false}
              inline={false}
              inputType="radio"
              labelAlign="right"
              labelText="pine apple"
              onChange={[Function]}
              standalone={true}
              value={3}
              wrapperClassName="col-xs-offset-2 col-xs-10"
            >
              <LayoutHelper
                FieldBase={
                  <input
                    checked={false}
                    onChange={[Function]}
                    type="radio"
                    value={3}
                  />
                }
                bsClass="radio"
                checked={false}
                inline={false}
                inputType="radio"
                labelAlign="right"
                labelText="pine apple"
                onChange={[Function]}
                singleCheckboxRadio={false}
                standalone={true}
                value={3}
                wrapperClassName="col-xs-offset-2 col-xs-10"
              >
                <div
                  className="radio"
                >
                  <Label
                    FieldBase={
                      <input
                        checked={false}
                        onChange={[Function]}
                        type="radio"
                        value={3}
                      />
                    }
                    className="checkbox-radio-label radio-wrapper"
                    inline={false}
                    inputType="radio"
                    labelAlign="right"
                    onChange={[Function]}
                    text="pine apple"
                    value={3}
                    wrapperClassName="col-xs-offset-2 col-xs-10"
                  >
                    <label
                      className="checkbox-radio-label radio-wrapper"
                    >
                      <input
                        checked={false}
                        onChange={[Function]}
                        type="radio"
                        value={3}
                      />
                      pine apple
                    </label>
                  </Label>
                </div>
              </LayoutHelper>
            </Radio>
          </div>
        </div>
      </div>
    </FormGroup>
  </LayoutHelper>
</RadioGroup>
`;

exports[`RadioGroup with props should render corrently 1`] = `
<RadioGroup
  bsClass="radio-group"
  inline={false}
  items={
    Array [
      Object {
        "labelText": "apple",
        "value": 1,
      },
      Object {
        "labelText": "pen",
        "value": 2,
      },
      Object {
        "labelText": "pine apple",
        "value": 3,
      },
    ]
  }
  labelAlign="right"
  onChange={[Function]}
  selectedValue={1}
  standalone={false}
>
  <LayoutHelper
    FieldBase={
      Array [
        <Radio
          bsClass="radio"
          checked={true}
          inline={false}
          inputType="radio"
          labelAlign="right"
          labelText="apple"
          onChange={[Function]}
          standalone={true}
          value={1}
          wrapperClassName="col-xs-offset-2 col-xs-10"
      />,
        <Radio
          bsClass="radio"
          checked={false}
          inline={false}
          inputType="radio"
          labelAlign="right"
          labelText="pen"
          onChange={[Function]}
          standalone={true}
          value={2}
          wrapperClassName="col-xs-offset-2 col-xs-10"
      />,
        <Radio
          bsClass="radio"
          checked={false}
          inline={false}
          inputType="radio"
          labelAlign="right"
          labelText="pine apple"
          onChange={[Function]}
          standalone={true}
          value={3}
          wrapperClassName="col-xs-offset-2 col-xs-10"
      />,
      ]
    }
    bsClass="radio-group"
    inline={false}
    items={
      Array [
        Object {
          "labelText": "apple",
          "value": 1,
        },
        Object {
          "labelText": "pen",
          "value": 2,
        },
        Object {
          "labelText": "pine apple",
          "value": 3,
        },
      ]
    }
    labelAlign="right"
    onChange={[Function]}
    selectedValue={1}
    singleCheckboxRadio={false}
    standalone={false}
    wrapperClassName="col-xs-10 col-xs-offset-2"
  >
    <FormGroup
      className=""
    >
      <div
        className="form-group"
      >
        <div
          className="col-xs-10 col-xs-offset-2"
        >
          <div
            className="radio-group"
          >
            <Radio
              bsClass="radio"
              checked={true}
              inline={false}
              inputType="radio"
              labelAlign="right"
              labelText="apple"
              onChange={[Function]}
              standalone={true}
              value={1}
              wrapperClassName="col-xs-offset-2 col-xs-10"
            >
              <LayoutHelper
                FieldBase={
                  <input
                    checked={true}
                    onChange={[Function]}
                    type="radio"
                    value={1}
                  />
                }
                bsClass="radio"
                checked={true}
                inline={false}
                inputType="radio"
                labelAlign="right"
                labelText="apple"
                onChange={[Function]}
                singleCheckboxRadio={false}
                standalone={true}
                value={1}
                wrapperClassName="col-xs-offset-2 col-xs-10"
              >
                <div
                  className="radio"
                >
                  <Label
                    FieldBase={
                      <input
                        checked={true}
                        onChange={[Function]}
                        type="radio"
                        value={1}
                      />
                    }
                    className="checked checkbox-radio-label radio-wrapper"
                    inline={false}
                    inputType="radio"
                    labelAlign="right"
                    onChange={[Function]}
                    text="apple"
                    value={1}
                    wrapperClassName="col-xs-offset-2 col-xs-10"
                  >
                    <label
                      className="checked checkbox-radio-label radio-wrapper"
                    >
                      <input
                        checked={true}
                        onChange={[Function]}
                        type="radio"
                        value={1}
                      />
                      apple
                    </label>
                  </Label>
                </div>
              </LayoutHelper>
            </Radio>
          </div>
          <div
            className="radio-group"
          >
            <Radio
              bsClass="radio"
              checked={false}
              inline={false}
              inputType="radio"
              labelAlign="right"
              labelText="pen"
              onChange={[Function]}
              standalone={true}
              value={2}
              wrapperClassName="col-xs-offset-2 col-xs-10"
            >
              <LayoutHelper
                FieldBase={
                  <input
                    checked={false}
                    onChange={[Function]}
                    type="radio"
                    value={2}
                  />
                }
                bsClass="radio"
                checked={false}
                inline={false}
                inputType="radio"
                labelAlign="right"
                labelText="pen"
                onChange={[Function]}
                singleCheckboxRadio={false}
                standalone={true}
                value={2}
                wrapperClassName="col-xs-offset-2 col-xs-10"
              >
                <div
                  className="radio"
                >
                  <Label
                    FieldBase={
                      <input
                        checked={false}
                        onChange={[Function]}
                        type="radio"
                        value={2}
                      />
                    }
                    className="checkbox-radio-label radio-wrapper"
                    inline={false}
                    inputType="radio"
                    labelAlign="right"
                    onChange={[Function]}
                    text="pen"
                    value={2}
                    wrapperClassName="col-xs-offset-2 col-xs-10"
                  >
                    <label
                      className="checkbox-radio-label radio-wrapper"
                    >
                      <input
                        checked={false}
                        onChange={[Function]}
                        type="radio"
                        value={2}
                      />
                      pen
                    </label>
                  </Label>
                </div>
              </LayoutHelper>
            </Radio>
          </div>
          <div
            className="radio-group"
          >
            <Radio
              bsClass="radio"
              checked={false}
              inline={false}
              inputType="radio"
              labelAlign="right"
              labelText="pine apple"
              onChange={[Function]}
              standalone={true}
              value={3}
              wrapperClassName="col-xs-offset-2 col-xs-10"
            >
              <LayoutHelper
                FieldBase={
                  <input
                    checked={false}
                    onChange={[Function]}
                    type="radio"
                    value={3}
                  />
                }
                bsClass="radio"
                checked={false}
                inline={false}
                inputType="radio"
                labelAlign="right"
                labelText="pine apple"
                onChange={[Function]}
                singleCheckboxRadio={false}
                standalone={true}
                value={3}
                wrapperClassName="col-xs-offset-2 col-xs-10"
              >
                <div
                  className="radio"
                >
                  <Label
                    FieldBase={
                      <input
                        checked={false}
                        onChange={[Function]}
                        type="radio"
                        value={3}
                      />
                    }
                    className="checkbox-radio-label radio-wrapper"
                    inline={false}
                    inputType="radio"
                    labelAlign="right"
                    onChange={[Function]}
                    text="pine apple"
                    value={3}
                    wrapperClassName="col-xs-offset-2 col-xs-10"
                  >
                    <label
                      className="checkbox-radio-label radio-wrapper"
                    >
                      <input
                        checked={false}
                        onChange={[Function]}
                        type="radio"
                        value={3}
                      />
                      pine apple
                    </label>
                  </Label>
                </div>
              </LayoutHelper>
            </Radio>
          </div>
        </div>
      </div>
    </FormGroup>
  </LayoutHelper>
</RadioGroup>
`;
